x/incentives/keeper/gauge_test.go

```diff
@@ -43,235 +43,69 @@ func (suite *KeeperTestSuite) TestNonExistentDenomGaugeCreation() {
-	suite.Require().NoError(err)
-	suite.Require().Equal(distrCoins, sdk.Coins{sdk.NewInt64Coin("stake", 200)})
-
-	// check active gauges
-	gauges = suite.App.IncentivesKeeper.GetActiveGauges(suite.Ctx)
-	suite.Require().Len(gauges, 1)
-
-	// check gauge ids by denom
-	gaugeIds := suite.App.IncentivesKeeper.GetAllGaugeIDsByDenom(suite.Ctx, "lptoken")
-	suite.Require().Len(gaugeIds, 1)
-
-	// check finished gauges
-	gauges = suite.App.IncentivesKeeper.GetFinishedGauges(suite.Ctx)
-	suite.Require().Len(gauges, 0)
-
-	// check rewards estimation
-	rewardsEst = suite.App.IncentivesKeeper.GetRewardsEst(suite.Ctx, lockOwner, []lockuptypes.PeriodLock{}, 100)
-	suite.Require().Equal(sdk.Coins(nil), rewardsEst)
-}
-
-func (suite *KeeperTestSuite) TestGaugesByDenom() {
-	// TODO: This is not a good test. We should refactor it to be table driven,
-	// specifying a list of gauges to define, and then the expected result of gauges by denom
-	testGaugeByDenom := func(isPerpetual bool) {
+	for _, tc := range testCases {
 		// test for module get gauges
 		suite.SetupTest()
 
 		// initial module gauges check
-		gaugeIds := suite.App.IncentivesKeeper.GetAllGaugeIDsByDenom(suite.Ctx, "lptoken")
-		suite.Require().Len(gaugeIds, 0)
+		gauges := suite.App.IncentivesKeeper.GetNotFinishedGauges(suite.Ctx)
+		suite.Require().Len(gauges, 0)
+		gaugeIdsByDenom := suite.App.IncentivesKeeper.GetAllGaugeIDsByDenom(suite.Ctx, "lptoken")
+		suite.Require().Len(gaugeIdsByDenom, 0)
 
 		// setup lock and gauge
-		_, gaugeID, _, startTime := suite.SetupLockAndGauge(isPerpetual)
+		lockOwner, gaugeID, coins, startTime := suite.SetupLockAndGauge(tc.isPerpetual)
 
 		// check gauges
-		gaugeIds = suite.App.IncentivesKeeper.GetAllGaugeIDsByDenom(suite.Ctx, "lptoken")
-		suite.Require().Len(gaugeIds, 1, "perpetual %b", isPerpetual)
-		suite.Require().Equal(gaugeIds[0], gaugeID)
+		gauges = suite.App.IncentivesKeeper.GetNotFinishedGauges(suite.Ctx)
+		suite.Require().Len(gauges, 1)
+		expectedGauge := types.Gauge{
+			Id:          gaugeID,
+			IsPerpetual: tc.isPerpetual,
+			DistributeTo: lockuptypes.QueryCondition{
+				LockQueryType: lockuptypes.ByDuration,
+				Denom:         "lptoken",
+				Duration:      time.Second,
+			},
+			Coins:             coins,
+			NumEpochsPaidOver: uint64(tc.expectedNumEpochsPaidOver),
+			FilledEpochs:      0,
+			DistributedCoins:  sdk.Coins{},
+			StartTime:         startTime,
+		}
+		suite.Require().Equal(expectedGauge.String(), gauges[0].String())
+
+		// check gauge ids by denom
+		gaugeIdsByDenom = suite.App.IncentivesKeeper.GetAllGaugeIDsByDenom(suite.Ctx, "lptoken")
+		suite.Require().Len(gaugeIdsByDenom, 1, "perpetual %b", tc.isPerpetual)
+		suite.Require().Equal(gaugeID, gaugeIdsByDenom[0])
+
+		// check rewards estimation
+		rewardsEst := suite.App.IncentivesKeeper.GetRewardsEst(suite.Ctx, lockOwner, []lockuptypes.PeriodLock{}, 100)
+		suite.Require().Equal(coins.String(), rewardsEst.String())
```

Comment:
I wonder if directly using `coins` here is appropriate given that it represents all the tokens in the gauge. Unless I'm missing something, the only reason this should be passing is because `SetupLockAndGauge` only sets up one  one lock (i.e. one account to distributes rewards to), and `coins / 1` = `coins` so it doesn't error. We might be able to increase coverage in general by setting this whole set of tests up with multiple gauges (or maybe even adding a "number of gauges" field to each test case).

The multi-lock case was originally covered in the non-perpetual gauge test, which set up many locks using `SetupManyLocks` and then directly filled the gauge using `SetupNewGauge` (see below). I'll leave it up to you to decide whether that kind of an implementation makes sense for the TDD version of these tests, but we should definitely be testing the scenario where multiple locks are being distributed to (for context, the correct expected value here would be `coins / n` where n is the number of locks).

The original implementation of this was inconsistent across perpetual and non-perp gauges so this is a completely understandable issue.

https://github.com/osmosis-labs/osmosis/blob/8f018a6c638b64912c1b22c7df04f84abe667ba0/x/incentives/keeper/gauge_test.go#L57

https://github.com/osmosis-labs/osmosis/blob/8f018a6c638b64912c1b22c7df04f84abe667ba0/x/incentives/keeper/gauge_test.go#L74

Reply:
Great catch, ended up adding `numLocks` as a column for our TDD